File: server/Homebank.Core/Domain/Entities/Transaction.cs
using Homebank.Core.Domain.Helpers;
using System;

namespace Homebank.Core.Domain.Entities
{
    public class Transaction : BaseEntity
    {
        private Category _category;

        public DateTime Date { get; private set; }
        public string Payee { get; private set; }
        public string Memo { get; private set; }
        public decimal OutFlow { get; private set; }
        public decimal Inflow { get; private set; }
        public bool IsBankTransaction { get; private set; }
        public bool IsInflowForBudgeting { get; private set; }
        public Category Category {
            get {
                if (IsInflowForBudgeting)
                {
                    return new Category("To be budgeted", new CategoryGroup("Inflow"));
                }

                return _category;
            }
            private set {
                _category = value;
            }
        }

        private Transaction()
        {
        }

        public Transaction(DateTime date, string payee, Category category, string memo, decimal outFlow, decimal inflow, bool isBankTransaction)
        {
            Guard.AgainstDefaultValue(date, "Must supply a valid date");

            if (inflow == default(decimal) && outFlow == default(decimal))
            {
                throw new ArgumentException("Must have at least an inflow or an outflow");
            }

            Date = date;
            Payee = payee ?? throw new ArgumentNullException(nameof(payee));
            Memo = memo ?? throw new ArgumentNullException(nameof(memo));
            OutFlow = outFlow;
            Inflow = inflow;
            IsBankTransaction = isBankTransaction;

            if (category?.Name.ToLower() == "inflow")
            {
                MarkAsTransactionForInflow(true);
            }
            else if (category != null)
            {
                AssignCategory(category);
            }
        }

        public void AssignCategory(Category category)
        {
            Guard.AgainstNull(category, nameof(category));

            IsInflowForBudgeting = false;
            Category = category;
        }

        public void MarkAsTransactionForInflow(bool isInflowForBudgeting)
        {
            if (isInflowForBudgeting)
            {
                Category = null;
            }

            IsInflowForBudgeting = isInflowForBudgeting;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            var transaction = obj as Transaction;

            return transaction != null
                    && transaction.Date == Date
                    && transaction.Payee == Payee
                    && transaction.Memo == Memo
                    && transaction.Inflow == Inflow
                    && transaction.OutFlow == OutFlow;
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hash = 17;
                hash = hash * 23 + Date.GetHashCode();
                hash = hash * 23 + Payee.GetHashCode();
                hash = hash * 23 + Memo.GetHashCode();
                hash = hash * 23 + Inflow.GetHashCode();
                hash = hash * 23 + OutFlow.GetHashCode();
                return hash;
            }
        }
    }
}

File: php-server/app/Features/Transactions/Transaction.php
<?php

namespace App\Features\Transactions;

use App\Core\Domain\Validations\Guard;
use App\Core\Domain\Validations\StringUtil;
use App\Exceptions\EntityValidationException;
use App\Features\Categories\Category;
use Assert\Assertion;
use DateTime;

class Transaction
{
    /** @var string */
    private $id;
    /** @var DateTime */
    private $date;
    /** @var string */
    private $payee;
    /** @var string */
    private $memo;
    /** @var string */
    private $outflow;
    /** @var string */
    private $inflow;
    /** @var bool */
    private $isBankTransaction;
    /** @var bool */
    private $isInflowForBudgeting;
    /** @var Category */
    private $category;

    public function __construct(
        DateTime $date,
        string $payee,
        Category $category,
        string $memo,
        string $outFlow,
        string $inFlow,
        bool $isInflowForBudgeting,
        bool $isBankTransaction)
    {
        Assertion::date($date);
        Assertion::notEmpty($payee); // https://github.com/beberlei/assert
        Assertion::notEmpty($memo);
        Assertion::numeric($outFlow);
        Assertion::numeric($inFlow);

        if (StringUtil::isEmpty($inFlow) && StringUtil::isEmpty($outFlow)) {
            throw new EntityValidationException('Must have at least an inflow or an outflow');
        }

        $this->date = $date;
        $this->payee = $payee;
        $this->memo = $memo;
        $this->outflow = $outFlow;
        $this->inflow = $inFlow;
        $this->isBankTransaction = $isBankTransaction;

        if ($isInflowForBudgeting) {
            $this->markTransactionForInFlow(true);
        } else if (!is_null($category)) {
            $this->assignCategory($category);
        }
    }

    private function markTransactionForInFlow(bool $isInFlowForBudgeting)
    {
        if ($this->isInflowForBudgeting) {
            $this->category = null;
        }

        $this->isInflowForBudgeting = $isInFlowForBudgeting;
    }

    private function assignCategory(Category $category)
    {
        Assertion::notNull($category);

        $this->isInflowForBudgeting = false;
        $this->category = $category;
    }
}
